### PR TITLE
fix(file source): fix message offset of opendal source

### PR DESCRIPTION
**Pull request by stdrc**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

#19654 changed the `OpendalReader::stream_read_object` to yield lines instead of byte buffers, but there's a accidental change of the semantics of the `SourceMessage::offset` for file source, which caused main-cron to [fail](https://buildkite.com/risingwavelabs/main-cron/builds/4047#0193a700-f6c8-4059-b6b1-d272a0ce4e82).

The change is that, previously, when `OpendalReader::stream_read_object` yield `SourceMessage`s of byte buffers, the `offset` is the STARTIG position of the BYTES, and after `split_stream` which splits these bytes by new-line, the new `offset` will be set to the ENDING position of each LINE, which is also the STARTING position of the NEXT LINE. #19654 changed the reader, and the offset is changed to always representing the STARTING position of each LINE, hence causing the error.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by stdrc** [comment]
* **#19721** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/19721?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/19721?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: src/connector/src/source/filesystem/opendal_source/opendal_reader.rs [hunk 1]
```diff
@@ -182,10 +182,16 @@ impl<Src: OpendalSource> OpendalReader<Src> {
...
+            // as the *message offset*, because we check whether a file is finished by comparing the
+            // message offset with the file size in `FsFetchExecutor::into_stream`. However, we must
+            // understand that this message offset is not semantically consistent with the offset of
+            // other source connectors.
+            let msg_offset = (offset + n_read).to_string();
             batch.push(SourceMessage {
                 key: None,
                 payload: Some(std::mem::take(&mut line_buf).into_bytes()),
-                offset: offset.to_string(),
+                offset: msg_offset,
```
Comment (xxchan) [inline src/connector/src/source/filesystem/opendal_source/opendal_reader.rs:194]: from my PoV, the meaning of "offset" for file source is too hard to understand and evaluate. (And my brain refuse to think about it.) Is it possible to change to sth like `(start,end)`?

**Review by xxchan — COMMENTED** [review]
Can we add some unit tests or e2e tests to demonstrate the behavior>?

File: src/connector/src/source/filesystem/opendal_source/opendal_reader.rs (hunk 1 above)
Comment (stdrc) [reply src/connector/src/source/filesystem/opendal_source/opendal_reader.rs:194]: > Is it possible to change to sth like `(start,end)`?

Agree, will consider doing this later.

**Comment by stdrc** [comment]
> Can we add some unit tests or e2e tests to demonstrate the behavior>?

Will need over 1k files to reproduce the bug. I think the s3 test in main-cron workflow is enough.

**Review by xxchan — APPROVED** [review]
rubber stamp

**Pull request merged**